app/assets/javascripts/components/buttons/generic-button-component.js

```diff
@@ -0,0 +1,41 @@
+ManageIQ.angular.app.component('genericButtonComponent', {
+  bindings: {
+    angularForm: '=',
+    modelCopy: '=',
+    model: '=',
+    redirectUrl: '=?',
+    newRecord: '<',
+    entity: '=?',
+    entityName: '=?',
+    onSaveClicked: '&?',
+    onAddClicked: '&?',
+  },
+  controllerAs: 'vm',
+  controller: genericButtonController,
+  templateUrl: '/static/buttons/generic_button_component.html.haml',
+});
+
+genericButtonController.$inject = ['miqService', '$window'];
+
+function genericButtonController(miqService) {
+  var vm = this;
+
+  vm.$onInit = function() {
+    vm.saveable = miqService.saveable;
+
+    vm.resetClicked = function() {
```

Comment:
If you ever appreciated the concept of code reusability, then you would appreciate the need for this component.

Like I said above, 75%-85% (or even more) forms would have this repeat code for Cancel and Reset ...something that we can easily address with this component. 
Not to mention the fact that we are also getting penalized by Code Climate which is dragging our GPA score down because of this redundancy.

Let me show you some data, and let that data speak for itself.

Here's a sampler -
```javascript
$git grep -A 4 resetClicked *.js

app/assets/javascripts/controllers/cloud_network/cloud_network_form_controller.js:  $scope.resetClicked = function() {
app/assets/javascripts/controllers/cloud_network/cloud_network_form_controller.js-    $scope.cloudNetworkModel = angular.copy( $scope.modelCopy );
app/assets/javascripts/controllers/cloud_network/cloud_network_form_controller.js-    $scope.angularForm.$setPristine(true);
app/assets/javascripts/controllers/cloud_network/cloud_network_form_controller.js-    miqService.miqFlash("warn", "All changes have been reset");
app/assets/javascripts/controllers/cloud_network/cloud_network_form_controller.js-  };
--
app/assets/javascripts/controllers/cloud_subnet/cloud_subnet_form_controller.js:  $scope.resetClicked = function() {
app/assets/javascripts/controllers/cloud_subnet/cloud_subnet_form_controller.js-    $scope.cloudSubnetModel = angular.copy( $scope.modelCopy );
app/assets/javascripts/controllers/cloud_subnet/cloud_subnet_form_controller.js-    $scope.angularForm.$setPristine(true);
app/assets/javascripts/controllers/cloud_subnet/cloud_subnet_form_controller.js-    miqService.miqFlash("warn", "All changes have been reset");
app/assets/javascripts/controllers/cloud_subnet/cloud_subnet_form_controller.js-  };
--
app/assets/javascripts/controllers/cloud_tenant/cloud_tenant_form_controller.js:  vm.resetClicked = function() {
app/assets/javascripts/controllers/cloud_tenant/cloud_tenant_form_controller.js-    vm.cloudTenantModel = angular.copy( vm.modelCopy );
app/assets/javascripts/controllers/cloud_tenant/cloud_tenant_form_controller.js-    $scope.angularForm.$setPristine(true);
app/assets/javascripts/controllers/cloud_tenant/cloud_tenant_form_controller.js-    miqService.miqFlash("warn", "All changes have been reset");
app/assets/javascripts/controllers/cloud_tenant/cloud_tenant_form_controller.js-  };
--
app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js:  vm.resetClicked = function(angularForm) {
app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js-    vm.cloudVolumeModel = angular.copy(vm.modelCopy);
app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js-    angularForm.$setPristine(true);
app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js-    miqService.miqFlash("warn", "All changes have been reset");
app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js-  };
--
app/assets/javascripts/controllers/configuration/time_profile_form_controller.js:  $scope.resetClicked = function() {
app/assets/javascripts/controllers/configuration/time_profile_form_controller.js-    vm.timeProfileModel = angular.copy( vm.modelCopy );
app/assets/javascripts/controllers/configuration/time_profile_form_controller.js-    $scope.angularForm.$setPristine(true);
app/assets/javascripts/controllers/configuration/time_profile_form_controller.js-    miqService.miqFlash("warn", __("All changes have been reset"));
app/assets/javascripts/controllers/configuration/time_profile_form_controller.js-  };
--
app/assets/javascripts/controllers/ansible_repository/repository_form_controller.js:  $scope.resetClicked = function() {
app/assets/javascripts/controllers/ansible_repository/repository_form_controller.js-    vm.repositoryModel = angular.copy( vm.modelCopy );
app/assets/javascripts/controllers/ansible_repository/repository_form_controller.js-    $scope.angularForm.$setPristine(true);
app/assets/javascripts/controllers/ansible_repository/repository_form_controller.js-    miqService.miqFlash('warn', __('All changes have been reset'));
app/assets/javascripts/controllers/ansible_repository/repository_form_controller.js-  };
--
app/assets/javascripts/controllers/auth_key_pair_cloud/auth_key_pair_cloud_controller.js:  vm.resetClicked = function() {
app/assets/javascripts/controllers/auth_key_pair_cloud/auth_key_pair_cloud_controller.js-    vm.keyPairModel = angular.copy( vm.modelCopy );
app/assets/javascripts/controllers/auth_key_pair_cloud/auth_key_pair_cloud_controller.js-    $scope.angularForm.$setPristine(true);
app/assets/javascripts/controllers/auth_key_pair_cloud/auth_key_pair_cloud_controller.js-    miqService.miqFlash("warn", __("All changes have been reset"));
app/assets/javascripts/controllers/auth_key_pair_cloud/auth_key_pair_cloud_controller.js-  };
--
```

Identical, mirror-image repeat code in all these controllers.
Why?
And this is not even the entire file list since it would be ridiculous to list *all* the `resetClicked`/`cancelClicked` instances here in this note.

Like I said...it's just a sampler... but you get the idea, right?

That should be some food for thought.